packages/env/lib/build-docker-compose-config.js

```diff
@@ -43,13 +43,52 @@ module.exports = function buildDockerComposeConfig( config ) {
 		...themeMounts,
 	];
 
-	const testsMounts = [
-		`${
-			config.coreSource ? config.coreSource.testsPath : 'tests-wordpress'
-		}:/var/www/html`,
-		...pluginMounts,
-		...themeMounts,
-	];
+	let testsMounts;
+	if ( config.coreSource ) {
+		testsMounts = [
+			// When using a local source for "core" we want to ensure two things:
+			//
+			// 1. That changes the user makes within the "core" directory are
+			//    served in both the development and tests environments.
+			// 2. That the development and tests environment use separate
+			//    databases and `wp-content/uploads`.
+			//
+			// To do this we copy the local "core" files ($wordpress) to a tests
+			// directory ($tests-wordpress) and instruct the tests environment
+			// to source its files like so:
+			//
+			// - wp-config.php        <- $tests-wordpress/wp-config.php
+			// - wp-config-sample.php <- $tests-wordpress/wp-config.php
+			// - wp-content           <- $tests-wordpress/wp-content
+			// - *                    <- $wordpress/*
+			//
+			// https://github.com/WordPress/gutenberg/issues/21164
+			`${ config.coreSource.testsPath }:/var/www/html`,
+			...fs
+				.readdirSync( config.coreSource.path )
```

Comment:
I don't really like how `https://github.com/WordPress/gutenberg/pull/21195` has become an "impure" function.

Reply:
Ah, this won't work when the core source isn't local as those files won't have been downloaded yet.